graph-builder/src/registry.rs

```diff
@@ -42,55 +42,148 @@ impl Into<cincinnati::Release> for Release {
+            r"^(?P<scheme>[a-z]+)(:/{2})(?P<host>([0-9a-zA-Z]+\.)*([0-9a-zA-Z]+)):(?P<port>[0-9]+)$"
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            let scheme = capture["scheme"].to_string();
+            return Ok(Registry {
+                insecure: Registry::insecure_scheme(&scheme)?,
+                scheme,
+                host: capture["host"].to_string(),
+                port: Some(
+                    capture["port"]
+                        .parse()
+                        .expect("could not parse port as a number"),
+                ),
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match scheme://h.o.s.t
+            r"^(?P<scheme>[a-z]+)(:/{2})(?P<host>([0-9a-zA-Z]+\.)*([0-9a-zA-Z]+))$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            let scheme = capture["scheme"].to_string();
+            return Ok(Registry {
+                insecure: Registry::insecure_scheme(&scheme)?,
+                scheme,
+                host: capture["host"].to_string(),
+                ..Default::default()
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match h.o.s.t:port
+            r"^(?P<host>([0-9a-zA-Z\-]+\.)+([0-9a-zA-Z]+)):(?P<port>[0-9]+)$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            return Ok(Registry {
+                host: capture["host"].to_string(),
+                port: Some(
+                    capture["port"]
+                        .parse()
+                        .expect("could not parse port as a number"),
+                ),
+                ..Default::default()
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match h.o.s.t
+            r"^(?P<host>([0-9a-zA-Z\-]+\.)*([0-9a-zA-Z]+))$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            return Ok(Registry {
+                host: capture["host"].to_string(),
+                ..Default::default()
+            });
+        };
+
+        bail!("unsupported registry format {}", src)
+    }
+
+    pub fn try_new(scheme: String, host: String, port: Option<u16>) -> Fallible<Self> {
+        Ok(Registry {
+            host,
+            port,
+            insecure: Self::insecure_scheme(&scheme)?,
+            scheme,
+        })
+    }
+
+    pub fn host_port_string(&self) -> String {
+        format!(
+            "{}{}",
+            self.host,
+            if let Some(port) = self.port {
+                format!(":{}", port)
+            } else {
+                "".to_string()
+            }
+        )
+    }
+
+    fn insecure_scheme(scheme: &str) -> Fallible<bool> {
+        match scheme {
+            "https" | "docker" => Ok(false),
```

Comment:
I don't think we want to allow a `docker` reference here, as the authority part is optional in some cases there (e.g. `busybox`).

Reply:
This won't be passed to dkregistry or elsewhere, it's just a safetycheck that the user doesn't expect us to fetch via `ftp` for example. I can remove it if it bothers you too much.